notifications_utils/postal_address.py

```diff
@@ -100,32 +121,57 @@ def has_valid_postcode(self):
 
     @property
     def has_valid_last_line(self):
-        return (self.allow_international_letters and self.international) or self.has_valid_postcode
+        return (
+            (self.allow_international_letters and self.international and not self.is_bfpo_address)
+            or self.has_valid_postcode
+            or (self.is_bfpo_address and not self.has_invalid_country_for_bfpo_address)
+        )
 
     @property
     def has_invalid_characters(self):
         return any(
             line.startswith(tuple(self.INVALID_CHARACTERS_AT_START_OF_ADDRESS_LINE)) for line in self.normalised_lines
         )
 
+    @property
+    def has_invalid_country_for_bfpo_address(self):
+        return self.international and self.is_bfpo_address
+
     @property
     def international(self):
         return self.postage != Postage.UK
 
+    @property
+    def is_bfpo_address(self):
+        return self._bfpo_number is not None
+
     @property
     def normalised(self):
         return "\n".join(self.normalised_lines)
 
     @property
     def normalised_lines(self):
+        if self.is_bfpo_address:
+            return self._lines_without_country_or_bfpo + [f"BFPO {self._bfpo_number}"]
 
         if self.international:
-            return self._lines_without_country + [self.country.canonical_name]
+            return self._lines_without_country_or_bfpo + [self.country.canonical_name]
+
+        if self.postcode:
+            return self._lines_without_country_or_bfpo[:-1] + [self.postcode]
+
+        return self._lines_without_country_or_bfpo
+
+    @property
+    def bfpo_normalised_lines(self):
```

Comment:
We also do something similar in `_parse_and_remove_bfpo()`, right?

Reply:
Yeah we do. (old) `_parse_and_remove_bfpo` actually edits the incoming raw lines. This works with the dynamic properties that power almost all of the other properties. I think it might be nice to remove some of the dynamic properties with one-off calculated values, effectively turning `PostalAddress` instances into immutable things. But I will leave that for the future.

Reply:
okee